Perf: faster retention time grouping using tree structures

JohanLassen opened the pull request:
Performance:
The retention time grouping function has been updated to reduce algorithmic complexity.

Computes retention time groups for >10,000 features in ~30 sec (before 2h). The algorithm is only implemented for retention time clustering and not for the group correlation module, as running times are already acceptable here (due to fewer comparisons).

Brief description of the algorithm:
(1) Calculate Manhattan distances
(2) Complete linkage clustering of distances
(3) Recursive search of clusters with intra-distances below maxDiff in top-down approach.
(3) Return groups from the identified clusters.

The algorithms gives a slightly different result but works as intended and passes the original tests.

JohanLassen committed:
perf: tree-based algorithm to group retention time.

JohanLassen committed:
performance improvements on retention time grouping

JohanLassen committed:
gitignore updated

JohanLassen committed:
Delete .Rproj.user directory

pushed by mistake

JohanLassen committed:
function documentation corrected

JohanLassen committed:
Resolved merge conflict in windowlayoutstate.pper

JohanLassen committed:
Merge pull request #1 from JohanLassen/tree-search

Tree search of retention time groups

sgibb reviewed (changes requested):
Thanks for this contribution.
Please remove the `*.Rproj` file from the repository and include a unit test.
Just for my background understanding: why do you use the manhattan instead of the euclidean distance?

JohanLassen committed:
requested changed made

JohanLassen committed:
Delete MsFeatures.Rproj

JohanLassen commented:
Thanks for the comments (It is my first pull request ever).
I made the suggested changes in my main branch that I am trying to pull. Does it sync to the pull request? 
I hope the testing is comprehensive enough. I only did it for the groupSimilarityMatrixTree function and not for the small helper function, get_cluster_members.

Manhattan and euclidean return the same distances for vectors (1D), but manhattan requires slightly less computations. The gain of doing this is absolutely neglectable compared to the time spend in the following steps :) So I don't mind changing to euclidean distances if that is easier.

jorainer commented:
Hi @JohanLassen ! Thanks for the PR and sorry for the delay - have been very busy. 

Regarding the PR question, yes, any changes you do and commit in your main branch will be added (and updated) to this PR. I will now check the code.

JohanLassen commented:
No worries- just happy that I might be able to actually contribute. It looks like I'm causing some troubles in git workflows. Let me know if I need to revise something.

jorainer commented:
I'll sort out the GHA problems - what puzzles me is that it looks like you replace the whole code in the *grouping-functions.R* file - but you only added two new functions and did not change any other code, right?

JohanLassen commented:
No, I changed the groupClosest function to use groupSimilarityMatrixTree which I wrote. Maybe it should be done completely separately? I thought the groupClosest was only used for retention times.

jorainer commented on R/grouping-functions.R:
```diff
@@ -1,335 +1,420 @@
+#' @importFrom stats hclust
+#' @family grouping operations
+groupSimilarityMatrixTree <- function(dists, maxDiff) {
+
```
please reduce the empty lines here and below

jorainer reviewed (changes requested):
Thanks for the contribution @JohanLassen ! I have some comments/change requests.

jorainer commented:
and @JohanLassen , can you please update/sync your fork with the main branch? If fixed the github actions.

JohanLassen committed:
groupClosest now support different grouping functions

JohanLassen committed:
Merge branch 'main' of github.com:JohanLassen/MsFeatures

JohanLassen commented:
Hi, sorry for the slow response. I corrected the comments and changed a bit in the groupClosest function so that it per default uses a matrix for the grouping function (FUN).

jorainer reviewed (changes requested):
Great! thanks for the update @JohanLassen ! I only have very minor change requests:

please add yourself as an contributor ("ctb") to the Description file. See e.g. [here](https://github.com/rformassspectrometry/SpectriPy/blob/main/DESCRIPTION#L41-L44) for an example. After that I'm fine to merge.

JohanLassen committed:
Description updated

JohanLassen committed:
Author description updated

JohanLassen committed:
Author description updated

JohanLassen committed:
noRd fixed

jorainer commented:
Thanks both for the updates. The two methods tackle different problems I believe - but it would be great to have both of them available. I'm not yet quite sure what the best way to group based on retention times is - and that very much depends also on the number of *things* to group - and their difference along RT... so, having different options might actually be great.

jorainer committed:
Merge branch 'main' into main

jorainer reviewed (approved):
Thanks Johan! I will merge and clean up a bit after the merge.

jorainer commented:
Hey Sebastian @sgibb - I will merge for now. I think your idea/approach with the algorithm from MALDIquant would be a great addition, but I would add that in a second step.

jorainer commented:
@JohanLassen I changed the default in `groupClosest()` from `FUN = groupSimilarityMatrixTree` to `FUN = groupSimilarityMatrix` to ensure that existing workflows don't get unexpectedly different results. `groupSimilarityMatrixTree()` and `groupSimilarityMatrix()` don't return identical results and I would first need to dig into that to understand why and what's *better*.

JohanLassen commented:
Of course. And thanks for the help :)
Here's a version of MALDIquant:::.binPeaks that use absolute thresholds - I changed as little as possible from the original code. I can implement it, but in my opinion we are good for now - I just wanted the speed up.

```
.grouperPairwise <- function(mass, intensities, samples, tolerance) {
  ## don't accept two or more peaks of the same sample
  if (anyDuplicated(samples)) {
    return(NA)
  }
  meanMass <- mean(mass)
  ## all peaks in range?
  if (any(as.matrix(dist(mass)) > tolerance)) {
    return(NA)
  }
  meanMass
}

mq_binpeaks <- function(x, tolerance) {
  p <- MALDIquant:::.binPeaks(x, tolerance = tolerance, samples = seq_along(x), grouper = .grouperPairwise)
  # hack to replace group means by consecutive group ids
  cumsum(!duplicated(p))
}
```
